tests/unit_tests/verRctNonSemanticsSimple.cpp

```diff
@@ -0,0 +1,248 @@
+// EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
+// MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL
+// THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+// SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
+// STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF
+// THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+#include <sstream>
+
+#define IN_UNIT_TESTS // to access Blockchain::expand_transaction_2
+
+#include "gtest/gtest.h"
+#include "unit_tests_utils.h"
+
+#include "cryptonote_basic/cryptonote_format_utils.h"
+#include "cryptonote_core/blockchain.h"
+#include "file_io_utils.h"
+#include "misc_log_ex.h"
+#include "ringct/rctSigs.h"
+
+/**
+ * @brief Used in test "cached_serialize_assumption" to test the do_serialize() assumption.
+ *
+ * The purpose of this macro is to assert that changes in rctSig's in memory correlate to changes in
+ * the serialized output of that rctSig. This assumes the existence of scoped variables
+ * "original_sig" and "original_sig_hex".
+ *
+ * @param fieldmodifyclause this should be an expression which modifies a rctSig field
+ */
+#define CACHED_SERIALIZATION_ASSUMPTION_SUBTEST(fieldmodifyclause)       \
+    do {                                                                 \
+        rct::rctSig modded_sig = original_sig;                           \
+        modded_sig.fieldmodifyclause;                                    \
+        const std::string mod_sig_hex = rct_sig_to_hex_hash(modded_sig); \
+        EXPECT_NE(original_sig_hex, mod_sig_hex);                        \
+    } while (0);                                                         \
+
+namespace
+{
+/**
+ * @brief Make rct::ctkey from hex string representation of destionation and mask
+ *
+ * @param dest_hex
+ * @param mask_hex
+ * @return rct::ctkey
+ */
+static rct::ctkey make_ctkey(const char* dest_hex, const char* mask_hex)
+{
+    rct::key dest;
+    rct::key mask;
+    CHECK_AND_ASSERT_THROW_MES(epee::from_hex::to_buffer(epee::as_mut_byte_span(dest), dest_hex), "dest bad hex: " << dest_hex);
+    CHECK_AND_ASSERT_THROW_MES(epee::from_hex::to_buffer(epee::as_mut_byte_span(mask), mask_hex), "mask bad hex: " << mask_hex);
+    return {dest, mask};
+}
+
+/**
+ * @brief Mimic how rctSig's are serialized and hashed in verRctNonSemanticsSimpleCached
+ *
+ * We convert the hash into a hex string though for ease of use.
+ * Check the body of verRctNonSemanticsSimpleCached to make sure this matches.
+ *
+ * @param rs ring signature
+ * @return std::string hash of serialization of the rctSig
+ */
+static std::string rct_sig_to_hex_hash(const rct::rctSig& rs)
```

Comment:
Afaict this function can just return a `crypto::hash` (`get_rct_sig_hash()`).

Reply:
Yeah it can, but I wanted to it be easier for humans to read. The hex representation is 1-to-1 so it shouldn't affect any outcomes.

Reply:
Especially useful in failure cases.

Reply:
In what circumstance is a human reading this and learning anything by looking at the bytes? Either the hashes match or they don't.

Reply:
Good point